pkg/plugins/runtime/k8s/controllers/gatewayapi/http_route_controller.go

```diff
@@ -0,0 +1,184 @@
+	kube_ctrl "sigs.k8s.io/controller-runtime"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayapi "sigs.k8s.io/gateway-api/apis/v1alpha2"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	"github.com/kumahq/kuma/pkg/core"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	"github.com/kumahq/kuma/pkg/core/resources/store"
+	k8s_common "github.com/kumahq/kuma/pkg/plugins/common/k8s"
+	"github.com/kumahq/kuma/pkg/plugins/resources/k8s"
+	k8s_util "github.com/kumahq/kuma/pkg/plugins/runtime/k8s/util"
+	util_k8s "github.com/kumahq/kuma/pkg/util/k8s"
+)
+
+// HTTPRouteReconciler reconciles a GatewayAPI object into Kuma-native objects
+type HTTPRouteReconciler struct {
+	kube_client.Client
+	Reader kube_client.Reader
+	Log    logr.Logger
+
+	Scheme          *kube_runtime.Scheme
+	Converter       k8s_common.Converter
+	SystemNamespace string
+	ResourceManager manager.ResourceManager
+}
+
+// Reconcile handles transforming a gateway-api HTTPRoute into a Kuma
+// GatewayRoute and managing the status of the gateway-api objects.
+func (r *HTTPRouteReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	httpRoute := &gatewayapi.HTTPRoute{}
+	if err := r.Get(ctx, req.NamespacedName, httpRoute); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			err = r.ResourceManager.Delete(ctx, core_mesh.NewGatewayRouteResource(), store.DeleteByKey(req.Name, req.Name))
+			if store.IsResourceNotFound(err) {
+				err = nil
+			}
+		}
+
+		return kube_ctrl.Result{}, err
+	}
+
+	// TODO check that gateways exist etc
+	// TODO set status on gapi resources
+
+	resource := core_mesh.NewGatewayRouteResource()
+
+	coreName := util_k8s.K8sNamespacedNameToCoreName(httpRoute.Name, httpRoute.Namespace)
+	mesh := k8s_util.MeshFor(httpRoute)
+
+	if err := r.ResourceManager.Get(ctx, resource, store.GetByKey(coreName, mesh)); err != nil {
+		if !store.IsResourceNotFound(err) {
+			return kube_ctrl.Result{}, err
+		}
+
+		resource.Spec, err = r.gapiToKumaRoute(httpRoute.Namespace, httpRoute)
+		if err != nil {
+			return kube_ctrl.Result{}, err
+		}
+
+		resource.SetMeta(&k8s.KubernetesMetaAdapter{ObjectMeta: metav1.ObjectMeta{
+			Name:      httpRoute.Name,
+			Namespace: httpRoute.Namespace,
+		}, Mesh: mesh})
+
+		err := r.ResourceManager.Create(ctx, resource, store.CreateByKey(coreName, mesh))
+
+		return kube_ctrl.Result{}, errors.Wrap(err, "could not create GatewayRoute resource")
+	}
+
+	var err error
+
+	resource.Spec, err = r.gapiToKumaRoute(httpRoute.Namespace, httpRoute)
+	if err != nil {
+		return kube_ctrl.Result{}, err
+	}
+
+	err = r.ResourceManager.Update(ctx, resource, store.ModifiedAt(core.Now()))
+
+	return kube_ctrl.Result{}, errors.Wrap(err, "could not update GatewayRoute resource")
+}
+
+func (r *HTTPRouteReconciler) gapiToKumaRoute(namespace string, route *gatewayapi.HTTPRoute) (*mesh_proto.GatewayRoute, error) {
```

Comment:
I think that what we should aim for here, and in the gateway controller, is a separate package that converts from kubernetes to gateway types. If we can do that with minimal infrastructure dependencies, then we can easily write tests to get good coverage the conversions.

 IMHO, this is OK to tackle in a separate PR.